From 3c3a541db6824849d330930efb7c30f89cf826fa Mon Sep 17 00:00:00 2001
From: John Ferlan <jferlan@redhat.com>
Date: Thu, 8 Aug 2013 15:46:48 -0400
Subject: [PATCH 06/48] spec: Replace the path to the tar.gz file

Commit '07adabc2' seems to have inadvertently removed the path
to the source ftp file added by commit '5412bf11'.

Signed-off-by: John Ferlan <jferlan@redhat.com>
---
 libvirt-cim.spec.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libvirt-cim.spec.in b/libvirt-cim.spec.in
index cd399b1..98f2a75 100644
--- a/libvirt-cim.spec.in
+++ b/libvirt-cim.spec.in
@@ -6,7 +6,7 @@ Version: @PACKAGE_VERSION@
 Release: 1%{?dist}%{?extra_release}
 License: LGPLv2+
 Group: Development/Libraries
-Source: libvirt-cim-%{version}.tar.gz
+Source: ftp://libvirt.org/libvirt-cim/libvirt-cim-%{version}.tar.gz
 BuildRoot: %{_tmppath}/%{name}-%{version}-%{release}-root
 URL: http://libvirt.org/CIM/
 Requires: libxml2 >= 2.6.0
-- 
1.8.5.3

